In [1]:
import mdtraj as md

In [2]:
from pathlib import Path

In [3]:
!pwd

/home/crystal/Desktop/L-allo-ile


In [4]:
traj = md.load_dcd('L_allo_prd.dcd', top='L_allo_clean.pdb')

In [5]:
traj

<mdtraj.Trajectory with 2000000 frames, 41 atoms, 2 residues, and unitcells at 0x7f22883e2950>

In [6]:
traj = md.Trajectory.superpose(traj, traj[0], frame=0, atom_indices=None, ref_atom_indices=None, parallel=True)

In [7]:
md.rmsd(traj, traj[0], frame=0, atom_indices=None, parallel=True, precentered=False)

array([0.        , 0.04252063, 0.04949549, ..., 0.2737564 , 0.26647475,
       0.24237768], dtype=float32)

In [8]:
md.rmsf(traj, traj[0], frame=0, atom_indices=None, parallel=True, precentered=False)

array([0.08387183, 0.13413525, 0.13497576, 0.13568191, 0.07279006,
       0.04638941, 0.0795862 , 0.04365111, 0.12399055, 0.13104221,
       0.18884365, 0.18943836, 0.19055068, 0.14729857, 0.10330804,
       0.1837379 , 0.16529453, 0.13538153, 0.18526007, 0.1838196 ,
       0.1838558 , 0.05435383, 0.08406099, 0.17308737, 0.08159116,
       0.12475886, 0.05008445, 0.12932082, 0.1598959 , 0.21767582,
       0.21832743, 0.2173537 , 0.23759948, 0.10142166, 0.18439786,
       0.17003568, 0.24030791, 0.1500413 , 0.20002337, 0.20095235,
       0.20054734], dtype=float32)

In [9]:
md.compute_phi(traj)

(array([[ 4, 21, 24, 23]]),
 array([[-1.4152273],
        [-1.3409921],
        [-1.5236   ],
        ...,
        [-1.281269 ],
        [-1.4981663],
        [-1.0591714]], dtype=float32))

In [10]:
md.compute_psi(traj)

(array([[ 0,  5,  4, 21]]),
 array([[2.1176953],
        [1.9228543],
        [2.0460203],
        ...,
        [2.7297847],
        [2.7743058],
        [2.5423226]], dtype=float32))

In [11]:
psi_indices, phi_indices = [ 0,  5,  4, 21], [ 4, 21, 24, 23]
angles = md.compute_dihedrals(traj, [phi_indices, psi_indices])
print(angles)

[[-1.4152273  2.1176953]
 [-1.3409921  1.9228543]
 [-1.5236     2.0460203]
 ...
 [-1.281269   2.7297847]
 [-1.4981663  2.7743058]
 [-1.0591714  2.5423226]]


In [12]:
angles.shape

(2000000, 2)

In [ ]:
from pylab import *
from math import pi

figure()
title('Dihedral Map: L-allo-Isoleucine dipeptide')
scatter(angles[:, 0], angles[:, 1], marker='x', c=traj.time)
cbar = colorbar()
cbar.set_label('Time [ps]')
xlabel(r'$\Phi$ Angle [radians]')
xlim(-pi, pi)
ylabel(r'$\Psi$ Angle [radians]')
ylim(-pi, pi)

fig, ax = plt.subplots(figsize=(6,6),tight_layout=True)
_,_,_,hist = ax.hist2d(angles[:, 0]*180/math.pi, angles[:, 1]*180/math.pi, bins=60, range=[[-180,180],[-180,180]])
ax.set_xlim([-180, 180])
ax.set_ylim([-180, 180])
ax.set_xlabel(r'$\Phi$ Angle', fontsize =15)
ax.set_ylabel(r'$\Psi$ Angle', fontsize =15)
ax.set_title('Dihedral Map: L-allo-Isoleucine dipeptide', fontsize =18)
ax.xaxis.set_ticks([-180,-120,-60,0,60,120,180])
ax.yaxis.set_ticks([-180,-120,-60,0,60,120,180])
plt.xticks(fontsize=13)
plt.yticks(fontsize=13)
plt.savefig('L-allo-Isoleucine dipeptide_dihedrals.png', dpi=600)

In [ ]:
import pyemma as pm

fig, ax = plt.subplots(figsize=(7,6),tight_layout=True)
fes = pm.plots.plot_free_energy(angles[:, 0]*180/math.pi, angles[:, 1]*180/math.pi, nbins=100, ax=ax)
ax.set_xlim([-180, 180])
ax.set_ylim([-180, 180])
ax.set_xlabel(r'$\Phi$ Angle (degree)', fontsize =15)
ax.set_ylabel(r'$\Psi$ Angle (degree)', fontsize =15)
ax.set_title('L-allo-Isoleucine dipeptide', fontsize =18)
ax.xaxis.set_ticks([-180,-120,-60,0,60,120,180])
ax.yaxis.set_ticks([-180,-120,-60,0,60,120,180])
plt.xticks(fontsize=13)
plt.yticks(fontsize=13)
plt.savefig('L-allo-Isoleucine dipeptide_freeenergy.png', dpi=600)
plt.show()

In [ ]:
fig, ax = plt.subplots(figsize=(7,6),tight_layout=True)
fes = pm.plots.plot_free_energy(angles[:, 0]*180/math.pi, angles[:, 1]*180/math.pi, nbins=100, ncontours=3, ax=ax)
ax.set_xlim([-180, 180])
ax.set_ylim([-180, 180])
ax.set_xlabel(r'$\Phi$ Angle (degree)', fontsize =15)
ax.set_ylabel(r'$\Psi$ Angle (degree)', fontsize =15)
ax.set_title('Contour Map: L-allo-Isoleucine dipeptide', fontsize =18)
ax.xaxis.set_ticks([-180,-120,-60,0,60,120,180])
ax.yaxis.set_ticks([-180,-120,-60,0,60,120,180])
plt.xticks(fontsize=13)
plt.yticks(fontsize=13)
plt.savefig('L-allo-Isoleucine dipeptide_contour.png', dpi=600)
plt.show()

In [14]:
int(angles.shape[0]/100)

20000

In [15]:
np.arange(0, int(angles.shape[0]/100)+1, 10)

array([    0,    10,    20, ..., 19980, 19990, 20000])

In [16]:
np.arange(0, int(angles.shape[0]/100)+1, 100).shape

(201,)

In [17]:
np.arange(0, int(angles.shape[0])+1, 10000).shape

(201,)

from pylab import *
from math import pi
psi = angles[:, 1].flatten()
phi = angles[:, 0].flatten()

fig, axes = plt.subplots(2,1, figsize=(20,8))

axes[0].plot(psi, lw=0.5)
axes[1].plot(phi, lw=0.5)
axes[0].set_title('L-allo-Isoleucine dipeptide', fontsize =18)
axes[0].set_ylabel(r'$\Psi$ Angle [radians]')
axes[1].set_ylabel(r'$\Phi$ Angle [radians]')
axes[1].set_xlabel('Time (ns)')
axes[0].set_xlabel('Time (ns)')

for ax in axes:
    ax.set_xticks(ticks=np.arange(0, int(angles.shape[0])+1, 10000), labels=np.arange(0, int(angles.shape[0]/100)+1, 100))
    ax.set_xlim([0, angles.shape[0]])

def find_transitions(s1_mask, s2_mask):
    s2s1_transition_ids, s1s2_transition_ids = [], []
    for i in tqdm(range(1, len(s1_mask)), desc=f'Finding transitions'):
        s1_ids = [k for k, x in enumerate(s1_mask[:i]) if x]
        s2_ids = [k for k, x in enumerate(s2_mask[:i]) if x]
        if len(s1_ids) == 0 or len(s2_ids) == 0:
            continue
        if s1_mask[i] and (s1_ids[-1] < s2_ids[-1]):
            s2s1_transition_ids.append([s2_ids[-1], i])
            with open('s2s1.txt', 'a') as f:
                # Write the list of strings to the file
                f.writelines(str(s2_ids[-1]) + ' ' + str(i) + '\n')
        if s2_mask[i] and (s1_ids[-1] > s2_ids[-1]):
            s1s2_transition_ids.append([s1_ids[-1], i])
            with open('s1s2.txt', 'a') as f:
                # Write the list of strings to the file
                f.writelines(str(s1_ids[-1]) + ' ' + str(i) + '\n')
    return s2s1_transition_ids, s1s2_transition_ids

def save_transitions(traj, s2s1_transition_ids, s1s2_transition_ids):
    if len(s2s1_transition_ids)>0:
        for no, ids in enumerate(s2s1_transition_ids):
            transition_traj = traj[ids[0]:ids[1]+1]
            transition_traj.superpose(transition_traj[0])
            transition_traj.save_pdb(md_transition_path.joinpath(f'traj_s2s1slowest{no}.pdb'))
    if len(s1s2_transition_ids)>0:
        for no, ids in enumerate(s1s2_transition_ids):
            transition_traj = traj[ids[0]:ids[1]+1]
            transition_traj.superpose(transition_traj[0])
            transition_traj.save_pdb(md_transition_path.joinpath(f'traj_s1s2slowest{no}.pdb'))
    return None

from tqdm import tqdm

from pathlib import Path

md_transition_path = Path('./')

find_transitions(phi>0.5, phi<-1.5)

s2s1_transition_ids = [[406252, 406255],
[479903, 479904],
[487583, 487584],
[838475, 838476],
[1156247, 1156248],
[1157515, 1157526],
[1171865, 1171866],
[1638263, 1638264],
[1745914, 1745915],
[1833049, 1833050],
[1898302, 1898309]]
s1s2_transition_ids =  [[133935, 133936],
[410399, 410415],
[479904, 479905],
[487584, 487585],
[838476, 838477],
[1156248, 1156249],
[1157883, 1157905],
[1171866, 1171867],
[1638264, 1638265],
[1747695, 1747702],
[1833050, 1833051],
[1898333, 1898346]]
save_transitions(traj, s2s1_transition_ids, s1s2_transition_ids)